### ISSUE TITLE
test(laguna): establish an external token-exact reference and validate Laguna S 2.1 NVFP4

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1765]
## Problem / Background

PR #1739 (issue #1347) landed the Laguna family and its report records two gaps the port could not close. `TECHNICAL_REPORTS/1739-laguna-compressed-tensors-nvfp4-experts-20260910.en.md:140`: the pinned mlx-lm checkout (0.31.3) carries no `laguna.py`, so #1347's token-exact greedy criterion is unmet and the only comparison available was fused versus unfused decode on `models/laguna-xs.2-4bit`, which diverges from about token 10 and reproduces on a Qwen3-30B-A3B control, making it the pre-existing fused affine decode kernel rather than the port. Line 142: Laguna S 2.1 was never downloaded, so its 48-layer path is unit-test-only. Line 144: `sqrtsoftplus` router scoring and `swa_attention_sink_enabled` sinks run only in tests, because no published checkpoint sets either.

## Current Behavior

`docs/supported-models.md:54` ends the Laguna entry with "Laguna S 2.1 was not available on the validation host". `src/models/laguna_tests.rs:233` (`router_sqrtsoftplus_matches_formula`) covers the scoring formula in isolation, but no test drives a forward pass with `moe_router_score_func` set, and the sink arm at `src/models/laguna_layers.rs:343` has no test at all (`laguna_tests.rs` contains no occurrence of `sink`). Both arms are confirmed unset in `models/laguna-xs-2.1-nvfp4/config.json` and in `poolside/Laguna-S-2.1-NVFP4`'s config.

## Proposed Solution

**Oracle.** "No oracle on either host" is true of mlx-lm only. Each published checkpoint ships its own modeling code: `models/laguna-xs-2.1-nvfp4/` already contains `modeling_laguna.py` and `configuration_laguna.py`, and `transformers` 5.4.0 is installed. That is the primary oracle, run out of band in Python and never in the request path, per the project's core principle that no Python runs in the request path. Its only missing prerequisite is `compressed_tensors`, which is not installed locally. A second oracle is reachable today: `references/mlx-vlm/mlx_vlm/models/laguna/language.py` (mlx-vlm 0.4.4) is a complete Laguna text implementation with the compressed-tensors unpack, router, expert stacking and sliding/full mix. Waiting for mlx-lm to ship `laguna.py` is a third option and blocks nothing.

**Compare on an NVFP4 checkpoint, not on XS.2.** Both NVFP4 configs declare the YaRN `attention_factor` a derived-factor oracle would compute anyway (XS 2.1 declares 1.3465735902799727 against `factor: 32`, S 2.1 declares 1.4852030263919618 against `factor: 128`, and `0.1 * ln(factor) + 1` reproduces both), so the oracle agrees there. `models/laguna-xs.2-4bit` declares `attention_factor: 1.0` where the derived value is 1.3466, so it diverges for the documented reason in report section 4.1 and is not a valid parity target. The mlx-vlm oracle inherits that limit: it imports `initialize_rope` from `mlx_lm.models.rope_utils` (`language.py:6`), which has no `attention_factor` handling, and its router (`language.py:38-66`) implements sigmoid and softmax only, so it cannot cover `sqrtsoftplus`.

**Reuse the trace format; do not write a second comparison tool.** Have the oracle emit teacher-forced traces in the TSV `examples/logit_trace.rs` already writes (`#`-prefixed metadata, then `chunk`, `pos`, `target`, `nll`, `top_ids`, `top_logits`, the last two comma-separated top-k lists, per `examples/logit_trace.rs:217` and `scripts/compare_logit_traces.py:52`), so `compare_logit_traces.py` reads it as one arm unmodified. Gate on disagreement at decided positions per `docs/benchmarks.md:407`, tracing at the width the code under test runs at, with the sixth positional argument supplying context.

**Rejected:** porting `laguna.py` into the pinned mlx-lm checkout (report line 193 lists it). It creates a vendored fork to maintain for no coverage the checkpoint's own modeling file lacks.

## Scope

**In scope:** the out-of-band oracle harness and its trace output; the `poolside/Laguna-S-2.1-NVFP4` download and smoke run; synthetic-config tests for the `sqrtsoftplus` forward path and the sink arm; the resulting `docs/supported-models.md:54` update.

**Out of scope:** the fused affine decode divergence (report section 5.3, reproduces on `models/qwen3-30b-a3b-4bit`, predates this branch, needs its own issue); the adapter route at `loading/config_backed.rs`, which report line 146 shows is a property of every `ConfigBacked` family rather than of Laguna; tensor parallel and batching (line 148); the DFlash drafter (#1351).

## Implementation Notes

- **Store name and size**: download as `models/laguna-s-2.1-nvfp4`, matching the sibling `models/laguna-xs-2.1-nvfp4`, with `./target/release/mlxcel download poolside/Laguna-S-2.1-NVFP4`. The HuggingFace API reports 61 files, 49 safetensors shards, 99.7 GB. `models` resolves to `/Users/inureyes/Development/models/mlx`, which has 163 GiB free today, so it fits with roughly 60 GiB left. Re-check free space before starting.
- **S 2.1 geometry to exercise**: 48 layers, `num_attention_heads_per_layer` of 48 on `full_attention` and 72 on `sliding_attention`, 8 KV heads, `head_dim` 128, 256 experts at top-10, `sliding_window` 512, YaRN `factor: 128` / `beta_fast: 32` / `original_max_position_embeddings: 8192` / `partial_rotary_factor: 0.5`, `max_position_embeddings` 1048576. Run on a host with 128 GB or more.
- **Reuse**: `examples/logit_trace.rs` and `scripts/compare_logit_traces.py` as they stand; its six positional arguments are `MODEL_DIR TEXT_FILE CHUNK_TOKENS MAX_CHUNKS TOPK PREFILL` (`examples/logit_trace.rs:119-135`).
- **Test placement**: `src/models/laguna_tests.rs`, beside `router_sqrtsoftplus_matches_formula` (line 233) and the forward tests (lines 713 and 758), built from a synthetic config in the style of `xs_config()` at line 70. Checkpoint-free, like the other 21 family tests.
- **Edge cases**: a divergence at position 0 is a tokenization or BOS mismatch rather than arithmetic, so check the `〈|EOS|〉` doubled-BOS rule (`MlxcelTokenizer::prompt_carries_bos`) before investigating kernels; an oracle on CPU in f32 against Metal f16 disagrees marginally, so record both arms' dtype; on an interrupted download, resume and confirm 49 shards before loading.
- **Constraint**: no Python in the request path. The harness lives under `scripts/` and is never imported by the runtime.

## Acceptance Criteria

- [x] An out-of-band oracle runs the checkpoint's own `modeling_laguna.py` (or mlx-vlm's Laguna) on `models/laguna-xs-2.1-nvfp4` and emits a teacher-forced trace that `scripts/compare_logit_traces.py` accepts unmodified. (PR #1779)
- [ ] First-64-token greedy identity is recorded against that oracle, or, where not exact, the first divergence position and the top-2 logit gap there, plus disagreement at decided positions per `docs/benchmarks.md:407`.
- [ ] The comparison runs on an NVFP4 checkpoint, and any run against `models/laguna-xs.2-4bit` attributes its `attention_factor` divergence to report section 4.1 rather than reporting it as a defect.
- [ ] `models/laguna-s-2.1-nvfp4` is in the store with 49 shards, and `mlxcel generate` on it gives finite logits, no NaN, and fluent Python.
- [x] A test drives a forward pass with `moe_router_score_func: "sqrtsoftplus"` set, beyond the formula-level coverage at `src/models/laguna_tests.rs:233`. (PR #1779)
- [x] A test covers the `swa_attention_sink_enabled` arm at `src/models/laguna_layers.rs:343`, which has none today. (PR #1779)
- [ ] `docs/supported-models.md:54` states what was established, replacing "Laguna S 2.1 was not available on the validation host".
- [ ] The workspace gate and lints pass.

## Verification

```bash
cargo test --workspace --profile test-fast --features metal,accelerate
cargo test --profile test-fast --features metal,accelerate --lib models::laguna
cargo clippy --workspace --all-targets -- -D warnings && cargo fmt --check

# S 2.1 fetch and smoke (128 GB or more; confirm free space first)
/bin/df -h /Users/inureyes/Development/models/mlx
./target/release/mlxcel download poolside/Laguna-S-2.1-NVFP4
./target/release/mlxcel generate -m models/laguna-s-2.1-nvfp4 -p "Write a Python retry wrapper with exponential backoff." -n 64 --no-chat-template --temp 0

# Oracle comparison on an NVFP4 checkpoint
cargo build --release --features metal,accelerate --example logit_trace
./target/release/examples/logit_trace models/laguna-xs-2.1-nvfp4 corpus.txt 8 60 8 512 > mlxcel.tsv
python3 scripts/<oracle-harness>.py models/laguna-xs-2.1-nvfp4 corpus.txt 8 60 8 512 > oracle.tsv
python3 scripts/compare_logit_traces.py mlxcel.tsv oracle.tsv
```

A pass is: the gate green, the S 2.1 smoke producing fluent Python with no NaN, and the comparison reporting zero disagreements at decided positions, or a recorded first-divergence position with its top-2 gap and an explanation.

## Blocking and interim state

Criterion (a) does not block on upstream mlx-lm. Its real prerequisite is running the checkpoint's own modeling code, which needs `compressed_tensors` installed (absent locally) or the GB10 (sm_121) CUDA host from report section 5.1. If neither is reachable, use the mlx-vlm oracle for the routing and attention paths, record that the YaRN `attention_factor` and `sqrtsoftplus` arms stayed uncovered by it, and leave `docs/supported-models.md:54` claiming only what the runs establish, the mitigation report line 193 already names.

Follow-up of #1347 (PR #1739).
